state.go

```diff
@@ -27,76 +31,804 @@ type State struct {
 	InterimTranscriptHash   []byte
 
 	// Shared secret state
-	Keys KeyScheduleEpoch
+	Keys keyScheduleEpoch
 
 	// Per-participant state
 	Index        leafIndex
 	IdentityPriv SignaturePrivateKey
+	scheme       SignatureScheme
 
 	// Cache of proposals and update secrets
 	// XXX: The map key for UpdateSecrets should actually be ProposalID, but that
 	// struct can't be used as a map key because it's not comparable (because
 	// slices are not comparable).  Instead, we use ProposalID.String()
 	PendingProposals []MLSPlaintext
 	UpdateSecrets    map[string][]byte
+
+	// local state to identify proposals being procesesed
+	// in the PendingProposals. Avoids linear loop to
+	// remove entries from PendingProposals.
+	processedProposals map[string]bool
 }
 
-func newEmptyState(groupID []byte, cs CipherSuite, leafPriv HPKEPrivateKey, cred Credential) (*State, error) {
-	// TODO
-	return nil, nil
+func newEmptyState(groupID []byte, cs CipherSuite, leafPriv HPKEPrivateKey, cred Credential) *State {
+	tree := newRatchetTree(cs)
+	tree.AddLeaf(0, &leafPriv.PublicKey, &cred)
+	secret := make([]byte, cs.newDigest().Size())
+	kse := newKeyScheduleEpoch(cs, 1, secret, []byte{})
+	s := &State{
+		CipherSuite:             cs,
+		GroupID:                 groupID,
+		Epoch:                   0,
+		Tree:                    *tree,
+		Keys:                    *kse,
+		Index:                   0,
+		IdentityPriv:            *cred.privateKey,
+		scheme:                  cred.Scheme(),
+		processedProposals:      map[string]bool{},
+		UpdateSecrets:           map[string][]byte{},
+		ConfirmedTranscriptHash: []byte{},
+		InterimTranscriptHash:   []byte{},
+	}
+	return s
 }
 
 func newJoinedState(ciks []ClientInitKey, welcome Welcome) (*State, error) {
-	// TODO
-	return nil, nil
+
+	suite := welcome.CipherSuite
+
+	s := new(State)
+	s.CipherSuite = suite
+	s.Tree = *newRatchetTree(suite)
+
+	var kp KeyPackage
+	var clientInitKey ClientInitKey
+	var found = false
+
+	// extract the keyPackage for init secret
+	for idx, cik := range ciks {
+		data, err := syntax.Marshal(cik)
+		if err != nil {
+			return nil, fmt.Errorf("mls.state: cik %d marshal failure %v", idx, err)
+		}
+		cikhash := welcome.CipherSuite.digest(data)
+		// parse the encryptedKeyPackage to find our right cik
+		for _, ekp := range welcome.EncryptedKeyPackages {
+			found = bytes.Equal(cikhash, ekp.ClientInitKeyHash)
+			if !found {
+				continue
+			}
```

Comment:
we should continue above since we didn't find match for our cikHash on line 97

Reply:
You need to make a decision -- either you keep searching if there's a problem with this entry, or you return on error.   Right now you do the latter, which I think is the right answer.  But in that case, you can break the loop here, since you're not going to keep searching.